### PR TITLE
[MXNET-454] Move distributed Docker cache from S3 to Docker Hub

### PR DESCRIPTION
**Pull request by marcoabreu**
## Description ##
At the moment, we download the entire cache from S3, amounting to up to 10GB of data. The problem here is that the download and the load parts have to run in sequence and especially the latter is quite slow (up to 4 minutes). This PR allows partial cache retrieval and parallelized loading.

I have added a test file to verify the functionality of full and partial caching. I will not add it to CI as it messes with the Docker Daemon and could have negative side effects. In general, just run it locally if you'd like to make a modification.

Things left:
- [x] Login to Docker Hub (using Secrets Manager)
- [x] Upload images to Docker Hub
- [x] Modify Jenkinsfile that handels cache generation
- [ ] Change Jenkins configuration to point to Docker Hub instead of S3 bucket
- [x] Update documentation 

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [x] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [x] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [x] Move from S3 to Docker Hub

### CODE REVIEW

File: ci/test_docker_cache.py [hunk 1]
```diff
@@ -0,0 +1,243 @@
...
+sys.path.append(os.path.dirname(__file__))
+import docker_cache
+import build as build_util
+
+DOCKERFILE_DIR = 'docker'
+DOCKER_REGISTRY_NAME = 'test_registry'
+DOCKER_REGISTRY_PORT = 5000
+DOCKER_REGISTRY_PATH = 'localhost:{}'.format(DOCKER_REGISTRY_PORT)
+
+class RedirectSubprocessOutput(object):
```
Comment (larroy) [inline ci/test_docker_cache.py]: Can you please describe what this class does

File: ci/test_docker_cache.py [hunk 2]
```diff
@@ -0,0 +1,243 @@
...
+        assert output.count('Running in') == expected_cache_miss_count, \
+            'Expected {} "Running in", got {}. Log:{}'.\
+                format(expected_cache_miss_count, output.count('Running in'), output)
+        assert output.count('Using cache') == expected_cache_hit_count, \
+            'Expected {} "Using cache", got {}. Log:{}'.\
+                format(expected_cache_hit_count, output.count('Using cache'), output)
+
+if __name__ == '__main__':
+    import nose2
+    nose2.main()
```
Comment (larroy) [inline ci/test_docker_cache.py]: add newline?

**Comment by larroy** [comment]
looks good, just minor comments

File: ci/build.py [hunk 3]
```diff
@@ -229,44 +228,55 @@ def script_name() -> str:
...
         platforms = get_platforms()
         logging.info("Building for all architectures: {}".format(platforms))
         logging.info("Artifacts will be produced in the build/ directory.")
         for platform in platforms:
-            if args.download_docker_cache:
-                import docker_cache
-                tag = get_docker_tag(platform)
-                logging.info('Docker cache download is enabled')
-                docker_cache.load_docker_cache(bucket_name=args.docker_cache_bucket, docker_tag=tag)
+            if args.docker_registry:
```
Comment (szha) [inline ci/build.py]: Seems to be repeating line 231?

File: ci/docker_cache.py [hunk 4]
```diff
@@ -28,237 +28,164 @@
...
-        else:
-            logging.debug('No cached image available for {}'.format(s3_object.key))
-    except botocore.exceptions.ClientError as e:
-        if e.response['Error']['Code'] == "404":
-            logging.debug('{} does not exist in S3 yet'.format(s3_object.key))
-        else:
-            raise
+    dockerhub_credentials = _get_dockerhub_credentials()
+    login_cmd = ['docker', 'login', '--username', dockerhub_credentials['username'], '--password',
+                 dockerhub_credentials['password']]
```
Comment (szha) [inline ci/docker_cache.py]: would anyone be able to update this file and print out the password in the log?

File: ci/docker_cache.py [hunk 5]
```diff
@@ -28,237 +28,164 @@
...
+    history_cmd = ['docker', 'history', '-q', docker_tag]
 
+    try:
+        image_ids_b = subprocess.check_output(history_cmd)
+        image_ids_str = image_ids_b.decode('utf-8').strip()
+        layer_ids = [id.strip() for id in image_ids_str.split('\n') if id != '<missing>']
 
-def _format_docker_cache_filepath(output_dir, docker_tag) -> str:
-    return os.path.join(output_dir, docker_tag.replace('/', '_') + '.tar')
+        delete_cmd = ['docker', 'image', 'rm', '--force']
```
Comment (szha) [inline ci/docker_cache.py]: deleting in a script looks scary... are there alternatives?

File: ci/docker_cache.py (hunk 4 above)
Comment (marcoabreu) [reply ci/docker_cache.py]: no, this only runs on restricted slaves which are not executing code from PRs. On regular slaves, this password is not accessible

File: ci/docker_cache.py (hunk 5 above)
Comment (marcoabreu) [reply ci/docker_cache.py]: Could you elaborate?

Comment (marcoabreu) [reply ci/docker_cache.py]: That's simply what you have to do if you want to clean the local docker cache

**Pull request merged**